### ISSUE TITLE
Add support for ConnectionString options in ReadCommand to configure the type mappings.

### ISSUE THREAD

**Issue by Rishabh-V** [googleapis/google-cloud-dotnet#8541]
This was caught and pointed by @jskeet while reviewing the code for making the type conversions configurable via connection string.

After the introduction of ConnectionString options to configure the CLR to Spanner and Spanner to CLR types, user should be able to configure the type mappings for Decimal and DateTime CLR types. However, if the user uses the following code:

```
using var cmd = connection.CreateReadCommand(
     "MyTable",
      ReadOptions.FromColumns("Col1", "Col2"),
      KeySet.FromParameters(new SpannerParameterCollection
      {
           new SpannerParameter { ParameterName = "Col1", Value = 3.14m}, //decimal
           new SpannerParameter { ParameterName = "Col2", Value = new DateTime(2022, 05, 31) }  // DateTime
       }));
using var reader = await cmd.ExecuteReaderAsync();
```

the connection string options are not honored.

**Comment by amanda-tarafa** [comment]
> I don't know how often users will actually create their own keys, so we could lock things down quite a bit.

You mean creating the keys from other than parameters? If that's the case, I think creating keys from other than parameters is a very common use case, as you can specify key ranges etc. Take a look at https://github.com/googleapis/google-cloud-dotnet/pull/7157#discussion_r706299751 from when the read API was added.

As for the API and the mechanism for converting the values, I say we try to mimic SpannerParameter as much as possible, as Key and KeySet are just a "different kind" of parameter and parameter collection respectively. This possibly means that we need to change Key and KeySet implementations?

**Comment by jskeet** [comment]
> This possibly means that we need to change Key and KeySet implementations?

Do you mean to make them do the conversion lazily (when they're used) instead of eagerly (when they're created)? That's what we do for SpannerParameter.

That would be a breaking change in terms of key equality, hash code and ToString representation, but maybe that's okay.

**Comment by jskeet** [comment]
One downside - currently it's definitely immutable. If we allow arbitrary objects to be specified, they might not be immutable.

We could potentially do *some* checking upfront, e.g. that this is *some* known CLR type for Spanner, and leave the rest. Arrays would still be a problem, mind you...

**Comment by amanda-tarafa** [comment]
> Do you mean to make them do the conversion lazily (when they're used) instead of eagerly (when they're created)? That's what we do for SpannerParameter.

Yes, exactly that.

I don't think we _need_ equality or even inmutability, which of course are always very nice to have. An then it would only be a breaking change for people using and changing mutable objects as part of the key, which is unlikely maybe? So I think I'm fine with it all.
And this is a problem with DbParameter and ParameterCollection, and we've actaully have some issues reported by users whoe reuse the parameters or the parameter collections while running commands in parallel.

But I'm fine with either proposal btw. The only note is that any solution needs to apply to keys and key sets built both from parameters and from "raw" values.

**Comment by jskeet** [comment]
Making the conversion lazy would be much cleaner in terms of not needing to change any API. I'm definitely happy with that if @Rishabh-V is. (I would suggest that we might want to do it "via" SpannerParameter for the implementation, to reduce the number of code paths that need to worry about this.)
